Servers/Administration/OJS.Servers.Administration/Controllers/ProblemResourcesController.cs

```diff
@@ -0,0 +1,186 @@
+                RouteValues = routeValues,
+            },
+        };
+
+        return base.Index();
+    }
+
+    public override Task<IActionResult> Create(IDictionary<string, string> complexId, string postEndpointName)
+        => base.Create(complexId, nameof(Create));
+
+    [HttpPost]
+    public Task<IActionResult> Create(IDictionary<string, string> entityDict, IFormFile file)
+        => base.PostCreate(entityDict, new FormFilesContainer(file));
+
+    public override Task<IActionResult> Edit(IDictionary<string, string> complexId, string postEndpointName)
+        => base.Edit(complexId, nameof(Edit));
+
+    [HttpPost]
+    public Task<IActionResult> Edit(IDictionary<string, string> entityDict, IFormFile file)
+        => base.PostEdit(entityDict, new FormFilesContainer(file));
+
+    public async Task<IActionResult> Download([FromQuery] IDictionary<string, string> complexId)
+    {
+        var id = this.GetEntityIdFromQuery(complexId);
+
+        var resource = await this.problemResourcesData.OneByIdTo<ProblemResourceDownloadServiceModel>(id);
+
+        await this.problemResourcesDownloadValidation
+            .GetValidationResult(resource)
+            .VerifyResult();
+
+        var file = resource?.File;
+
+        if (file == null)
+        {
+            return this.NotFound();
+        }
+
+        var contentType = this.contentTypes.GetByFileExtension(resource!.FileExtension);
+        var fileName = $"Resource-{resource.Id}-{resource.ProblemName.Replace(" ", string.Empty)}.{resource.FileExtension}";
+
+        return this.File(file, contentType, fileName);
+    }
+
+    protected override IEnumerable<Func<ProblemResource, ProblemResource, AdminActionContext, ValidatorResult>>
+        EntityValidators
+        => this.problemResourceValidatorsFactory.GetValidators();
+
+    protected override IEnumerable<Func<ProblemResource, ProblemResource, AdminActionContext, Task<ValidatorResult>>>
+        AsyncEntityValidators
+        => this.problemResourceValidatorsFactory.GetAsyncValidators();
+
+    protected override IEnumerable<GridAction> CustomActions
+        => new[]
+        {
+            new GridAction { Action = nameof(Download) }
+        };
+
+    protected override object GetDefaultRouteValuesForPostEntityFormRedirect(
+        ProblemResource newEntity)
+        => new Dictionary<string, string>
+        {
+            { ProblemIdKey, newEntity.ProblemId.ToString() },
+        };
+
+    protected override async Task<IEnumerable<FormControlViewModel>> GenerateFormControlsAsync(
+        ProblemResource entity,
+        EntityAction action,
+        IDictionary<string, string> entityDict,
+        IDictionary<string, Expression<Func<object, bool>>> complexOptionFilters)
+    {
+        var formControls = await base.GenerateFormControlsAsync(entity, action, entityDict, complexOptionFilters)
+            .ToListAsync();
+
+        var problemId = entityDict.TryGetEntityId<Problem>();
```

Comment:
`TryGetEntityById` is not a good name, based on the .Net convensions.
Better name is `GetEntityIdOrDefault()`